phoenix-ios/phoenix-ios/views/transactions/TxHistoryExporter.swift

```diff
@@ -0,0 +1,570 @@
+							do {
+								try FileManager.default.removeItem(at: tmpFileUrl)
+								log.debug("Deleted tmp file: \(tmpFileUrl)")
+							} catch {
+								log.error("Error deleting tmp file: \(error)")
+							}
+						}
+						shareUrl = nil
+					}
+				}
+			}
+		)
+	}
+	
+	private func fetchOldestPaymentDate() {
+		log.trace("fetchOldestPaymentDate()")
+		
+		let databaseManager = Biz.business.databaseManager
+		databaseManager.paymentsDb { (result: SqlitePaymentsDb?, _) in
+			
+			assertMainThread()
+			guard let paymentsDb = result else {
+				return
+			}
+			
+			paymentsDb.getOldestCompletedDate { (millis: KotlinLong?, _) in
+				
+				if let oldestDate = self.millisToDate(millis) {
+					log.debug("oldestDate = \(oldestDate)")
+					startDate = oldestDate
+				}
+			}
+		}
+	}
+	
+	private func sanitizeStartDate() -> Int64 {
+		
+		let startOfDay = Calendar.current.startOfDay(for: startDate)
+		return Int64(startOfDay.timeIntervalSince1970 * 1_000)
+	}
+	
+	private func sanitizeEndDate() -> Int64 {
+		
+		var components = DateComponents()
+		components.day = 1
+		components.second = -1
+		
+		let endOfDay = Calendar.current.date(byAdding: components, to: endDate)!
+		return Int64(endOfDay.timeIntervalSince1970 * 1_000)
+	}
+	
+	
+	private func refreshPaymentCount() {
+		log.trace("refreshPaymentCount()")
+		
+		let startMillis = sanitizeStartDate()
+		let endMillis = sanitizeEndDate()
+		
+		let databaseManager = Biz.business.databaseManager
+		databaseManager.paymentsDb { (result: SqlitePaymentsDb?, _) in
+			
+			assertMainThread()
+			guard let paymentsDb = result else {
+				return
+			}
+			
+			paymentsDb.listRangePaymentsCount(
+				startDate: startMillis,
+				endDate: endMillis
+			) { (result: KotlinLong?, _) in
+				
+				let count = result?.intValue ?? 0
+				if (startMillis == self.sanitizeStartDate()) && (endMillis == self.sanitizeEndDate()) {
+					paymentCount = count
+				} else {
+					// result no longer matches UI; user changed dates
+				}
+			}
+		}
+	}
+	
+	private func millisToDate(_ millis: KotlinLong?) -> Date? {
+			
+		if let millis = millis {
+			let seconds: TimeInterval = millis.doubleValue / Double(1_000)
+			return Date(timeIntervalSince1970: seconds)
+		} else {
+			return nil
+		}
+	}
+	
+	@MainActor
+	private func exportingFailed() {
+		log.trace("exportingFailed()")
+		
+		isExporting = false
+		toast.pop(
+			NSLocalizedString("Exporting Failed", comment: "TxHistoryExporter"),
+			colorScheme: colorScheme.opposite,
+			style: .chrome,
+			duration: 15.0,
+			alignment: .middle,
+			showCloseButton: true
+		)
+	}
+	
+	@MainActor
+	private func exportingFinished(_ tempFile: URL) {
+		log.trace("exportingFinished()")
+		
+		isExporting = false
+		shareUrl = tempFile
+	}
+	
+	// --------------------------------------------------
+	// MARK: Exporting
+	// --------------------------------------------------
+	
+	@MainActor
+	private func asyncExport() async {
```

Comment:
I think the logic that writes the CSV strings should be moved to the shared module, in Kotlin. If we do that, how easy would it be to used from Swift? (while keeping the part writing the CSV string to file/disk in Swift)

Reply:
Done in [5887637](https://github.com/ACINQ/phoenix/pull/337/commits/5887637bfd95bd07bdd1671fa9d0b38a1063c6b1)